src/lib/support/StringCircularBuffer.cpp

```diff
@@ -0,0 +1,151 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <support/StringCircularBuffer.h>
+
+#include <limits>
+#include <nlassert.h>
+#include <string.h>
+
+#include <support/CodeUtils.h>
+
+namespace chip {
+
+std::size_t StringCircularBuffer::Advance(std::size_t dataLocation, size_t amount) const
+{
+    dataLocation += amount;
+    if (dataLocation >= mCapacity)
+        dataLocation -= mCapacity;
+    return dataLocation;
+}
+
+void StringCircularBuffer::Read(uint8_t * dest, size_t length, size_t offset) const
+{
+    VerifyOrDie(Size() >= offset + length);
```

Comment:
Please document that we don't need to worry about overflow because this is not a public method and we make sure we never call it with a length/offset that would overflow on addition?

Though that begs the question of why we are doing this check anyway.